File: android/app/src/main/java/com/tastyhub/android/ui/screens/home/HomeViewModel.kt
package com.tastyhub.android.ui.screens.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.tastyhub.android.data.model.RestaurantOut
import com.tastyhub.android.data.remote.TastyHubApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

private const val PAGE_SIZE = 24

data class HomeUiState(
    val restaurants: List<RestaurantOut> = emptyList(),
    val query: String = "",
    val isLoading: Boolean = false,
    val isLoadingMore: Boolean = false,
    val canLoadMore: Boolean = true,
    val error: String? = null,
)

class HomeViewModel(private val api: TastyHubApi) : ViewModel() {
    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState

    private var searchJob: Job? = null

    init {
        loadRestaurants(reset = true)
    }

    fun onQueryChange(query: String) {
        _uiState.value = _uiState.value.copy(query = query)
        searchJob?.cancel()
        searchJob = viewModelScope.launch {
            delay(350)
            loadRestaurants(reset = true)
        }
    }

    fun loadMore() {
        val state = _uiState.value
        if (state.isLoading || state.isLoadingMore || !state.canLoadMore) return
        loadRestaurants(reset = false)
    }

    fun retry() = loadRestaurants(reset = true)

    private fun loadRestaurants(reset: Boolean) {
        viewModelScope.launch {
            val state = _uiState.value
            _uiState.value = state.copy(isLoading = reset, isLoadingMore = !reset, error = null)
            val offset = if (reset) 0 else state.restaurants.size
            val search = state.query.ifBlank { null }
            runCatching { api.getRestaurants(search = search, limit = PAGE_SIZE, offset = offset) }
                .onSuccess { page ->
                    _uiState.value = _uiState.value.copy(
                        restaurants = if (reset) page else _uiState.value.restaurants + page,
                        isLoading = false,
                        isLoadingMore = false,
                        canLoadMore = page.size == PAGE_SIZE,
                    )
                }
                .onFailure {
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        isLoadingMore = false,
                        error = it.message ?: "Failed to load restaurants",
                    )
                }
        }
    }
}

class HomeViewModelFactory(private val api: TastyHubApi) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T = HomeViewModel(api) as T
}

File: ios/TastyHub/Screens/Home/HomeViewModel.swift
import Foundation

@MainActor
final class HomeViewModel: ObservableObject {
    @Published var query = "" {
        didSet { scheduleSearch() }
    }
    @Published private(set) var restaurants: [RestaurantOut] = []
    @Published private(set) var isLoading = false
    @Published private(set) var isLoadingMore = false
    @Published private(set) var errorMessage: String?

    private var canLoadMore = true
    private var searchTask: Task<Void, Never>?
    private let api = APIClient.shared
    private let pageSize = 24

    init() {
        Task { await load(reset: true) }
    }

    private func scheduleSearch() {
        searchTask?.cancel()
        searchTask = Task {
            try? await Task.sleep(nanoseconds: 350_000_000)
            guard !Task.isCancelled else { return }
            await load(reset: true)
        }
    }

    func retry() {
        Task { await load(reset: true) }
    }

    func loadMore() {
        guard !isLoading, !isLoadingMore, canLoadMore else { return }
        Task { await load(reset: false) }
    }

    private func load(reset: Bool) async {
        if reset { isLoading = true } else { isLoadingMore = true }
        errorMessage = nil
        let offset = reset ? 0 : restaurants.count
        do {
            let page = try await api.getRestaurants(search: query.isEmpty ? nil : query, limit: pageSize, offset: offset)
            restaurants = reset ? page : restaurants + page
            canLoadMore = page.count == pageSize
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to load restaurants"
        }
        isLoading = false
        isLoadingMore = false
    }
}
